src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs

```diff
@@ -45,6 +45,23 @@ public static ExpressionSyntax WalkDownParentheses(this ExpressionSyntax express
             return expression;
         }
 
+        public static ExpressionSyntax WalkDownParenthesesAndExclamations(this ExpressionSyntax expression)
+        {
+            while (expression.IsKind(SyntaxKind.ParenthesizedExpression) || expression.IsKind(SyntaxKind.SuppressNullableWarningExpression))
```

Comment:
this seems very unlike how we do anything else.  Specifically it does not seem to behave like how we handle any of:

`.`, `->` or `?.`

This feels very wrong, unless there is a super good reason for doing it this way.

Reply:
if I were you, I would just point him the existing code and tell him we had existing code and doing the similar thing in this pattern rather than telling him all these things and still not pointing to the code you are referring.

and I am not sure what you meant by it is wrong. 
this is existing code
http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/Extensions/ExpressionSyntaxExtensions.cs,38

Reply:
> and I am not sure what you meant by it is wrong.

I mean, I would expect the code to handle `!.` would be very similar in nature to the existing code we have to handle one of these forms:  `.`, `->`, `?.`, or `::`.  It's not.  Something is very fishy there.

Reply:
> if I were you, I would just point him the existing code

My VS is crashing non-stop right now.  And i don't really have time to figure out why.  It woudl take a fair amount of time to go find this with other tools, and I'm not really being payed for this.  So i gave my feedback that i thought this doesn't at all seem right, and I'm happy to engage further depending on how much time I have and how well my VS is working.

Reply:
From my discussion with @AlekseyTs , we found that `!.` quite differs from `?.`. Syntactically, `!.` is the same as `++.` or `--.`. Semantically, at https://github.com/dotnet/roslyn/blob/0f5f87379090fbe906fb6c27209d970dd0175321/src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationService.cs#L101

`!.` goes as `SyntaxKind.SimpleMemberAccessExpression`, while `?.` is considered as `SyntaxKind.MemberBindingExpression`. 

So, the idea is to ignore `?` the same as parenthesis.

Reply:
> So, the idea is to ignore ? the same as parenthesis.


I think you mean "ignore `!` the same as parentheses".

however, i don't know why you'd need to ignore it in the first place.  Say you have: ```foo!.```

We'll have the `.` with the `foo!` on the left.  We'll ask what the type of that expression is.  That should give us the correct type from the semantic model.  So we should just be able to list the members of that type just fine, right?

Reply:
Having `foo!.` and ignoring `!` the same way as `()`, we should provide the same completion for `foo!.` as it would be for `foo.`. This is the idea.

Reply:
@ivanbasov  Perhaps we should chat on gitter.  My point is that it does not seem as if any work would be necessayr, and this simply indicates some sort of issue with the semantic model.  Hitting `dot` should say "give me the items for the type of the thing prior to the dot".  In tihs case, the thing prior to the dot is simply `foo!`.  And it should be possible to ask for that guy's type and then list hte members.  There should be no need to unwrap the `!`.